### PR TITLE
CI: add ASAN job (build and check-all)

### PR DESCRIPTION
**Pull request by nurmukhametov**
Add CI job to build ISPC with the address sanitizer (ASAN). 

According to [this](https://github.com/google/sanitizers/wiki/AddressSanitizerLeakSanitizer), ASAN includes also the leak sanitizer (LSAN). At the moment, LSAN is turned off by `ASAN_OPTIONS=detect_leaks=0`, so only ASAN operates. It could find memory errors like double-free, use-after-free, stack/heap corruptions.

### CODE REVIEW

File: .github/workflows/ispc-ci-san.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    - name: Install dependencies
+      run: |
+        .github/workflows/scripts/install-build-deps.sh
+
+    - name: Check environment
+      run: |
+        ./check_env.py
+        which -a clang
+        which -a gcc
```
Comment (dbabokin) [inline .github/workflows/ispc-ci-san.yml:31]: Let's document that gcc is used because the clang is built without `compiler-rt` libraries.

Comment (nurmukhametov) [reply .github/workflows/ispc-ci-san.yml:31]: I have documented that.

File: .github/workflows/ispc-ci-san.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        .github/workflows/scripts/install-build-deps.sh
+
+    # We will use system gcc until we rebuild LLVM with compiler-rt.
+    - name: Check environment
+      run: |
+        ./check_env.py
+        which -a gcc
+        cat /proc/cpuinfo
+
+    - name: Build package
```
Comment (dbabokin) [inline .github/workflows/ispc-ci-san.yml]: `Build package` is misleading name here, as only config is done. I suggest keeping the name, but doing package build here as well.

File: .github/workflows/ispc-ci-san.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    - name: Check environment
+      run: |
+        ./check_env.py
+        which -a gcc
+        cat /proc/cpuinfo
+
+    # Address sanitizer includes leak sanitizer but we don't want it to report
+    # all leaks at the moment because there are quite a lot if them.
+    # So, we disable leaks detection to have clean report.
+    - name: Run ASAN
```
Comment (dbabokin) [inline .github/workflows/ispc-ci-san.yml:37]: Let's have 3 different steps here for clarity / readability of the run:
- config and build ISPC with ASAN enabled
- run ASAN on LIT tests
- run ASAN on runtime tests

**Review by dbabokin — CHANGES_REQUESTED** [review]
We also need to clean up GCC warnings - not as part of this PR, I would suggest having this as a separate job (with fixed GCC version and `-Werror` enabled)

File: .github/workflows/ispc-ci-san.yml (hunk 3 above)
Comment (nurmukhametov) [reply .github/workflows/ispc-ci-san.yml]: Fixed, test run is here https://github.com/nurmukhametov/ispc/actions/runs/5535409990/jobs/10101667972

Comment (nurmukhametov) [reply .github/workflows/ispc-ci-san.yml]: Example of failed job on double-free error is here https://github.com/nurmukhametov/ispc/actions/runs/5535736851/jobs/10102438055

**Review by dbabokin — APPROVED** [review]
I've push the changes that I assumed in the review feedback. If the look good, please go forward with the merge.

**Comment by nurmukhametov** [comment]
> I've push the changes that I assumed in the review feedback. If the look good, please go forward with the merge.

It looks good (test run is [here](https://github.com/nurmukhametov/ispc/actions/runs/5541689886/jobs/10115419740)).

**Pull request merged**